### ISSUE TITLE
Adding Feature: AI-based summarization task for WhatsApp chats with UI viewer

### ISSUE THREAD

**Issue by mbichara** [sepinf-inc/IPED#2641]
This feature adds a new AISummarizationTask that generates summaries with LLMs (Large Language Models) for item contents. Initially implemented for WhatsApp chats.

The task sends the WhatsApp chat contents to a remote service (AI middleware service) and stores returned textual summaries on each item’s extra attributes.

In the analysis interface, when the user clicks on on item that has "summaries" attributes, these are rendered on a "Summary" tab near Preview tab on bottom right. The Summary tab should be hidden otherwise.

The motivation of this feature is to provide quick, high-level insights over long conversations to accelerate analysis and triage.

The AISummarizationTask runs after the AudioTranscriptionTask, allowing to create summaries of conversations including transcripted audio messages. 

The idea is also to later extend the feature to other item types, chats and documents.
These summaries can also be leaveraged in the future for desiging a RAG (Retrieval-Augmented Generation) achitecture with LLMs.

**Comment by mbichara** [comment]
I was struggling with the UI logic. I created an SummaryViewer, and added it to the ViewerController class. It works as I thought it should, but I think the changes there are not ideal (most problably not). Glad if anyone could review and help with that.

Also feel free to make any suggestions to the feature regarding code or architecture-related.

**Comment by marcosammoura** [comment]
This will be a very useful feature, @mbichara. I intend to colaborate with you on this task and, if possible, work together on the server-side AI algorithms.

**Comment by felipecampanini** [comment]
Good evening, everyone!

We’ve done something similar (though a bit simpler) here at SETEC/SP.
The Python task generates a summary of all chats based on the prompt stored in the configuration file, and saves this summary as metadata for the chat in IPED.

The server is running the Llama 3.1 B model, quantized to 4 bits (that’s the maximum we managed to run with our current setup: 1 HP Z8 and 4 GTX 1660 SUPER GPUs).

One thing we noticed during use is that the model’s context window is quite small, so we had to split the analysis every 20,000 characters. Ideally, the entire chat should be processed at once — or at least with fewer chunks — to avoid losing context.

I’m sharing the code below

Special thanks to @wladimirleite for helping with the development.

ChatAnalyzerTask.py
```python
# Chat analysis using a local network hugging face model via a Flask server
# =============
# Configuration
# Enabling: enableChatAnalysis in IPEDConfig.txt
# Config file: ChatAnalysis.txt
# =============
# Version 0.3
# Changelog 
#   26/08/2025 - (@wladimirleite) Changed from "chat:analysis:score" to "Communication:analysis:score", return to using meta_data.set, bookmark bug fix
#   20/08/2025 - Changed meta_data.set to .add to fix Jep TypeError
#   19/08/2025 - First version
# =============
import logging
import json
import requests

API_URL = None
SYSTEM_PROMPT = None
enableProp = "enableChatAnalysis"
enabled = False
bookmarkCreated = False

class ChatAnalyzerTask:

    def isEnabled(self):
        return enabled
    
    def finish(self):
        global bookmarkCreated
        if bookmarkCreated:
            return

        query = "Communication\\:analysis\\:score: [0 TO *]"
        queryScore = "Communication\\:analysis\\:score: [750 TO *]"

        try:
            searcher.setQuery(query)
            ids = searcher.search().getIds()
            
            if len(ids) > 0:
                bookmarkId = ipedCase.getBookmarks().newBookmark("IA: Chats Analisados")
                ipedCase.getBookmarks().setBookmarkComment(bookmarkId, f"{len(ids)} chats foram analisados pela IA.")
                ipedCase.getBookmarks().addBookmark(ids, bookmarkId)
                ipedCase.getBookmarks().saveState(True)

            searcher.setQuery(queryScore)
            ids = searcher.search().getIds()
            
            if len(ids) > 0:
                bookmarkId = ipedCase.getBookmarks().newBookmark("IA: Chats Analisados de maior relevância")
                ipedCase.getBookmarks().setBookmarkComment(bookmarkId, f"{len(ids)} chats foram analisados pela IA com maior relevância, baseados no score.")
                ipedCase.getBookmarks().addBookmark(ids, bookmarkId)
                ipedCase.getBookmarks().saveState(True)

            bookmarkCreated = True

        except Exception as e:
            logging.error(f"[ERROR] Falha ao criar bookmark no método finish: {e}")

    def getConfigurables(self):
        from iped.engine.config import EnableTaskProperty, DefaultTaskPropertiesConfig
        return [
            EnableTaskProperty(enableProp), 
            DefaultTaskPropertiesConfig(enableProp, "ChatAnalysis.txt")
        ]

    def init(self, configuration):
        global enabled, API_URL, SYSTEM_PROMPT
        enabled = configuration.getEnableTaskProperty(enableProp)
        if not enabled:
            return
        
        taskConfig = configuration.getTaskConfigurable("ChatAnalysis.txt")
        extraProps = taskConfig.getConfiguration()
        
        API_URL = extraProps.getProperty("apiUrl")
        SYSTEM_PROMPT = extraProps.getProperty("prompt")

        if not API_URL or not SYSTEM_PROMPT:
            logging.error("[ERROR] ChatAnalysisTask configuration is incomplete. Check apiUrl and prompt in ChatAnalysis.txt")
            enabled = False
        else:
            logging.info("[INFO] ChatAnalysisTask Initialized Successfully.")

    def process(self, item):
        try:
            media_type = item.getMediaType().toString()
            
            ALLOWED_MEDIA_TYPES = ["application/x-whatsapp-chat",
             "application/x-ufed-chat-whatsapp",
             "application/x-ufed-chat-preview-whatsapp",
             "application/x-ufed-chat-preview",
             "application/x-telegram-chat",
             "application/x-ufed-chat-telegram",
             "application/x-ufed-chat-preview-telegram",
             "application/x-ufed-chat-preview-facebook",
             "application/x-ufed-chat-preview-instagram",
             "application/x-discord-chat",
             "application/x-ufed-html-chats",
             "application/x-ufed-chats-txt",
             "application/x-ufed-chat",
             "application/x-ufed-chat-preview"]
            
            if media_type not in ALLOWED_MEDIA_TYPES:
                return

            chat_content = "\n".join([line for line in item.getParsedTextCache().splitlines() if line.strip()])

            if not chat_content:
                return

            #logging.warning(f"[DEBUG] ITEM IS ALLOWED: {item.getName()}. Calling API.")
            analysis_result = self.analyze_chat_via_api(chat_content)
            #logging.warning(f"[DEBUG] API RESULT RECEIVED: {analysis_result}")

            meta_data = item.getMetadata()
            if "error" in analysis_result:
                meta_data.add("Communication:analysis:error", str(analysis_result["error"]))
                return

            average_score = analysis_result.get("average_score", 0)
            response_text = analysis_result.get("response_text", "")

            meta_data.set("Communication:analysis:score", str(average_score))
            meta_data.set("Communication:analysis:response", str(response_text))
            
            #logging.warning(f"[DEBUG] Finished analyzing item {item.getName()}. METADATA SET.")
        except Exception as e:
            logging.error(f"[ERROR] Unhandled exception in process method: {e}")

    def analyze_chat_via_api(self, chat_text):
        result = {}
        payload = {"prompt": SYSTEM_PROMPT, "text": chat_text}
        
        try:
            response = requests.post(API_URL, json=payload, timeout=8300)
            response.raise_for_status()
            result = response.json()
        except requests.exceptions.RequestException as e:
            result["error"] = f"Error connecting to local Llama server: {e}"
        except Exception as e:
            result["error"] = f"An unexpected error occurred: {e}"
        return result

```

**llama_server.py**
```python
import torch
from transformers import pipeline, BitsAndBytesConfig
from flask import Flask, request, jsonify
import logging
import math
import threading
import json

logging.basicConfig(filename='llamaserver.log', filemode='a', format='%(asctime)s - %(levelname)s - %(message)s', level=logging.INFO)
CHUNK_SIZE = 20000 # To avoid OutOfMemory due to window size;
analysis_lock = threading.Lock()
app = Flask(__name__)
pipe = None

# Model loading
try:
    logging.info("Checking device availability (GPU/CPU)...")
    
    pipeline_args = {
        "model_kwargs": {
            "torch_dtype": torch.bfloat16
        }
    }

    if torch.cuda.is_available():
        device_count = torch.cuda.device_count()
        logging.info(f"Success! {device_count} CUDA-compatible GPU(s) found:")
        for i in range(device_count):
            logging.info(f"  - GPU {i}: {torch.cuda.get_device_name(i)}")
        
        quantization_config = BitsAndBytesConfig(load_in_4bit=True)
        
        pipeline_args['device_map'] = 'auto'
        pipeline_args['model_kwargs']['quantization_config'] = quantization_config
        logging.info("Model will be distributed and quantized (4-bit) automatically among the found GPUs.")

    else:
        logging.warning("No CUDA-compatible GPUs were found. The model will be loaded onto the CPU (this will be slower).")
        pipeline_args['device'] = 'cpu'
    
    logging.info("Loading model Llama 3.1 8B. This may take a few minutes...")    
    MODEL_ID = "meta-llama/Meta-Llama-3.1-8B-Instruct"
    pipe = pipeline(
        "text-generation",
        model=MODEL_ID,
        **pipeline_args
    )
    logging.info("Model loaded successfully!")
    logging.info(f"First layer device of the model: {pipe.model.device}")

except Exception as e:
    logging.error(f"Failed to load model:{e}")

@app.route('/analyze', methods=['POST'])
def analyze_chat():
    if pipe is None:
        return jsonify({"error": "Falha ao carregar o modelo. Verifique o console do servidor."}), 500

    logging.info("New request received, waiting in queue (if necessary)...")
    analysis_lock.acquire()
    logging.info("Lock acquired. Starting analysis.")
    
    try:
        data = request.get_json()
        if not data or 'text' not in data or 'prompt' not in data:
            return jsonify({"error": "Payload inválido. 'text' e 'prompt' são obrigatórios."}), 400

        chat_text = data['text']
        system_prompt = data['prompt']
        
        # Instructs the model to always respond in valid JSON format;
        json_instruction = '\n\nResponda SEMPRE em formato JSON com as chaves "score" (um número de 0 a 1000 indicando a relevância do chat para a investigação) e "resposta" (sua análise em texto).'
        system_prompt_with_json = system_prompt + json_instruction

        total_chunks = math.ceil(len(chat_text) / CHUNK_SIZE)
        all_scores = []
        all_text_responses = []
        logging.info(f"The text will be divided into {total_chunks} part(s)")

        for i in range(total_chunks):
            chunk_num = i + 1
            logging.info(f"Processing {chunk_num}/{total_chunks}...")
            
            start_index = i * CHUNK_SIZE
            end_index = start_index + CHUNK_SIZE
            chunk_text = chat_text[start_index:end_index]

            messages = [
                {"role": "system", "content": system_prompt_with_json},
                {"role": "user", "content": chunk_text},
            ]

            outputs = pipe(
                messages,
                max_new_tokens=1024,
                do_sample=True,
                temperature=0.6,
                top_p=0.9,
            )
            
            response_text_chunk = outputs[0]["generated_text"][-1]['content']
            
            try:
                # Limpa a resposta para extrair apenas o JSON
                json_part = response_text_chunk[response_text_chunk.find('{'):response_text_chunk.rfind('}')+1]
                parsed_json = json.loads(json_part)
                score = parsed_json.get("score", 0)
                resposta = parsed_json.get("resposta", "")
                
                if isinstance(score, (int, float)):
                    all_scores.append(score)
                
                formatted_response = f"[PARTE {chunk_num}/{total_chunks}]:\n{resposta}"
                all_text_responses.append(formatted_response)

            except (json.JSONDecodeError, AttributeError) as e:
                logging.error(f"Unable to extract JSON from chunk {chunk_num}. AI response: {response_text_chunk}. Error: {e}")
                all_text_responses.append(f"[PARTE {chunk_num}/{total_chunks} - ERRO DE PARSING]:\n{response_text_chunk}")

            logging.info(f"Analysis of chunk {chunk_num} complete.")

        average_score = 0
        if all_scores: 
            average_score = sum(all_scores) / len(all_scores)
            
        logging.info(f"Complete consolidated analysis. Average score: {average_score:.2f}. Sending response.")
        
        # Retorna um JSON com a média do score e o texto concatenado
        # return jsonify({"average_score": average_score, "response_text": concatenated_text})
        return jsonify({"average_score": average_score, "response_text": all_text_responses})
    finally:
        analysis_lock.release()
        logging.info("Analysis completed and lock released. Server ready for the next request.")

if __name__ == '__main__':
    if pipe is not None:
        #Inserir o IP da máquina aqui
        app.run(host='0.0.0.0', port=5000)
    else:
        logging.error("The server cannot be started because the template failed to load.")
```


ChatAnalysis.txt
```
# URL for the local Flask server endpoint
apiUrl=http://0.0.0.0:5000/analyze

criarMarcadores=false

# The system prompt that will guide the AI's analysis.
prompt=Inserir prompt aqui 
```
IPEDConfig.txt
```
# Enable or disable the Chat Analysis task
enableChatAnalysis=true
```

**Comment by mbichara** [comment]
Hi @felipecampanini! Just saw you message now, very nice, thanks for sharing. I liked the idea of getting an score as output, it's something I will look at. 

Lately I've been working on server side, and some benchmarking of the VLLM inference engine. What I found interesting is that VLLM has prefix caching feature (https://docs.vllm.ai/en/v0.10.2/features/automatic_prefix_caching.html), so subsequest requests with same input are much faster, as I tested.

This could be interesting for us as we could, for example, perform LLM summarization of a chunk, and then right after perform other LLM requests for analysis related to investigation (classification and/or scoring) of the same chunk, with the performance benefit.

**Comment by mbichara** [comment]
Hi, everyone!

I've had good progress on this new feature, to the point that now I believe some collaboration will be very valuable.

Some Notes: As it is, the task sends the chat contents (WhatsApp/UFED chats) to a remote service (AI middleware service). The server uses an LLM to generate and return textual summaries of parts (chunks) of the chats. These are stored in each item’s extra attributes (ai:summaries), which are then rendered on a "Summary" tab.

Following the idea and code of @felipecampanini and @wladimirleite, I added a chat analysis functionality to the task. This allows setting questions and the respective question attribute names that will store a score from 0 to 999 (e.g., ai:analysis:bankingFraudScore). The good thing, as I've been testing, is that prefix caching seems to work nicely on the VLLM server, as I mentioned in my last message, boosting performance.

The main parts that require more work are:

-Summary tab viewer: The IPED UI part is still tricky for me. The quick additions I made to ViewerController.java are causing laggy and weird behavior. I saw that @wladimirleite programmed this; if you could give some help there, it would be very helpful.

-Suggestions on showing question scores nicely: I really liked the AI Panel feature implemented by the colleagues. I think it would be nice to show the chat analysis results in the AIPanel, like it's done with the NudityScore. Example: BankingFraudScore - Very low, low, high, very high.

-Workload and general testing/review

I am open to any suggestions or improvements. If anyone could help on this, I already have a development AI server running. Just let me know, and I can assist in getting the task running on it.

**Comment by wladimirleite** [comment]
@mbichara great work!
That will be a very useful feature!

I will be on vacation in the next two weeks, but I can try to help when I return.

I haven't tried to run it, but from what you described, I may have a few suggestions in mind. In particular, I am not sure if a new tab in the viewer would be the best approach. I have a couple of alternatives in mind, but I would need to look into them in more detail.

**Comment by mbichara** [comment]
Hi @wladimirleite !

Thank you very much. Sure, we talk later, enjoy your vacations!

**Comment by wladimirleite** [comment]
Hi @mbichara! 
I was able to process real evidence using the summarization task. 
The results are really great! 
Awesome job!!!

I have some minor suggestions regarding the user interface. 
Most are really minor, but can still make a difference in daily usage.

One of my main questions when I read the issue description and code changes, before running it, was whether the "Summary" panel makes sense, or if it should show the summary at the beginning of the chat view (in a similar way GMail shows an AI-generated summary at the top of the e-mail thread itself). 
However, I observed when I processed real evidence that the summary generated for longer chats contains more detail (and characters). So, placing it "inside" the chat itself wouldn't work as I expected. One advantage of the dedicated summary tab is that other types of files could be summarized in the future (not only chats).

I still have to look at the "questions" part. 
Then I will post the suggestions I have in the next few days.

**Comment by wladimirleite** [comment]
> Then I will post the suggestions I have in the next few days.

I will post my comments/suggestions in the PR, as they are related to the implementation.

**Comment by lfcnassif** [comment]
I just updated the Title to reflect current implementation of #2646

**Comment by wladimirleite** [comment]
> I just updated the Title to reflect current implementation of [#2646](https://github.com/sepinf-inc/IPED/pull/2646)

@mbichara can provide a more accurate feedback, but I think the current implementation is creating entries on the AI panel based on scores for user defined questions, instead of bookmarks.

**Comment by lfcnassif** [comment]
Thanks @wladimirleite, I just updated the Title again.
